### PR TITLE
Skip computation of much of last layer & unused logits during prompt eval / large N

### PR DESCRIPTION
**Pull request by ochafik**
**(Bit-rotten since multimodel support &) Superseded by https://github.com/ggerganov/llama.cpp/pull/6122 🎉**

**TL;DR**: the speedup is modest but consistent on CPU (~4% for 7B model, 15-25% for TinyLlamas models on M2 Mac) for a rather simple change

During prompt eval (w/ N > 1), a lot of what is done at the last layer is wasted unless we want all the logits out: it's fine to focus on the last token once that layer finished writing to the kv-cache.

Concretely, this means we can drop all but the last column of `cur` and `inpSA` after writing to the KV cache at the last layer, unless `logits_all` is set (I've moved `tmpq` & `Qcur` to right after that narrowing reshape).

Some notes on performance:

- Squeezing **eval time** will only make a difference when the prompt is much larger than the generation.
- The more layers, the least benefits
    - Zero benefit for 70B models
    - 4-5% faster **eval** of 7B models on CPU, and 3% on Metal (M2 Max)
    - TinyLlamas 15M, 42M, 110M models get CPU eval speedups of 15%, 20%, 25% ([test instructions](https://gist.github.com/ochafik/d75ef799d7288ddd3238bcb2b7631a66)) (FWIW I initially played with this idea on the lean llama2.c's codebase, _where it's easier to understand what is being skipped than w/ llama.cpp's tensor batch formulation_: https://github.com/karpathy/llama2.c/pull/253)
- Inconclusive results on CUDA, not sure how to get stable results w/o dedicated hardware (tried on [this Colab](https://colab.research.google.com/gist/ochafik/664f78524d8650a7b71b322d1b55d98d/llama-cpp-skip-unused-logits-computations.ipynb))
- This won't make perplexity computation any faster (all logits are needed there)
- Technically could skip more operations when the eval is chunked in batches (since only the last logit of the last chunk is needed), but there's vastly diminishing returns here

To test this branch:

```bash
git clone https://github.com/ochafik/llama.cpp ochafik-llama.cpp
cd ochafik-llama.cpp
git checkout skip-unused-2 && git pull

cmake -B ../build-skip . && ( cd ../build-skip && make -j main ) 
cmake -B ../build-noskip . -DLLAMA_SKIP_UNUSED_LOGITS=0  && ( cd ../build-noskip && make -j main ) 

# Edit to point to your model's path
hyperfine --warmup 1 \
  -L build noskip,skip \
  '../build-{build}/bin/main -m llama-2-7b-chat.ggmlv3.q2_K.bin --temp 0 -n 1 -f prompts/reason-act.txt -ngl 0'
# set -ngl 1 for Metal / CUDA
```

### CODE REVIEW

**Comment by KerfuffleV2** [comment]
This doesn't seem compatible with #1087 (the ROCM port). I just get garbage output trying to evaluate a model. Doesn't seem to affect the perplexity tool (as expected).

I tested with CPU only and CLBlast - seems to work fine with those. (Only tested evaluation.)

**Comment by ochafik** [comment]
@KerfuffleV2 Thanks a lot for testing it out! I've now disabled this when LLAMA_USE_HIPBLAS is defined, but after glancing through that PR I don't see any reason why it wouldn't work, although I have very low understanding of anything CUDA- or ROCm-related.

@ggerganov Do you have existing tricks to compare intermediate outputs between backends? (some way to force synchronous evaluation of each operation maybe, e.g. w/ its own 1-node graph & compute call; maybe appending each tensor to a binary file then reading it / comparing it on the fly?). That would allow spotting at which exact op garbage starts appearing (could potentially reveal an underlying bug in the ROCm code if it's the only backend causing issues).

If there's interest in shipping this PR (which is arguably borderline useful given low single-digit speedup) I'll try and rent a cloud ROCm instance to debug 🤗

**Comment by KerfuffleV2** [comment]
@SlyEcho Any idea of what could be going on here? https://github.com/ggerganov/llama.cpp/pull/2700#issuecomment-1689548127

**Comment by SlyEcho** [comment]
Should not be related to AMD, will have to test CUDA as well.

I suspect something about the shape of the tensors is not suited to our optimized kernels.

File: llama.cpp [hunk 1]
```diff
@@ -32,6 +32,12 @@
 #endif
 #endif
 
+// TODO: Fix unused logit skipping crashes on ROCm
+// (see https://github.com/ggerganov/llama.cpp/pull/2700#issuecomment-1689548127)
+#ifndef LLAMA_USE_HIPBLAS
+#define LLAMA_SKIP_UNUSED_LOGITS
+#endif
+
```
Comment (SlyEcho) [inline llama.cpp:40]: I suggest removing this and instead make LLAMA_SKIP_UNUSED_LOGITS configurable from make/CMake.

Comment (KerfuffleV2) [reply llama.cpp:40]: Guess it depends on whether this works for everything _except_ ROCM.

Comment (ochafik) [reply llama.cpp:64]: Yes I was hoping for this to be a default mini optimization that always works *except* in some temporary cases, but I'll prepare a cmake / makefile update.

I've tested the output on CUDA again ([updated colab](https://colab.research.google.com/gist/ochafik/664f78524d8650a7b71b322d1b55d98d/llama-cpp-skip-unused-logits-computations.ipynb#scrollTo=aU88u6VBTQwX&line=2&uniqifier=1)), included a branch where I've merged the ROCm changes and the CUDA backend appears to skip well on T4 w/ and w/o the ROCm changes.

Comment (KerfuffleV2) [reply llama.cpp:64]: > included a branch where I've merged the ROCm changes and the CUDA backend appears to skip well on T4 w/ and w/o the ROCm changes.

I'm not sure I fully understand this. Do you mean you're saying the changes like:

```patch
- const float dall = x[i].dm.x;
+ const float dall = __low2half(x[i].dm);
```

aren't the issue when running with CUDA? That's certainly useful information since it would mean the non-ROCM-specific changes don't cause a problem for CUDA.

Most likely the problem I originally described is specific to actually using ROCM on an AMD GPU.

**Comment by Engininja2** [comment]
I tried the collab with some -ngl values smaller than 33 and they produce garbage for output.

**Comment by ardfork** [comment]
Yes, I wanted to check what the problem was on ROCm, but it's the same if fully offloading it work but if offloading a few layers it doesn't.

**Comment by ochafik** [comment]
@Engininja2 @ardfork Thanks, great to be able to repro the issue!

@KerfuffleV2 @SlyEcho This PR's definitely broken on CUDA independently of the ROCm changes __when not all layers are offloaded to the GPU__. I don't understand the offload_func_* logic yet, reckon I'll need to update it.

**Comment by SlyEcho** [comment]
I suspect it's got something to do with copying memory between device and host.
I think @JohannesGaessler and @slaren would know better.

File: llama.cpp [hunk 2]
```diff
@@ -2323,6 +2320,37 @@ static struct ggml_cgraph * llm_build_llama(
                 ggml_build_forward_expand(gf, ggml_cpy(ctx0, Vcur, v));
             }
 
+#ifdef LLAMA_SKIP_UNUSED_LOGITS
+            if (il == n_layer - 1 && !lctx.logits_all)
+            {
```
Comment (JohannesGaessler) [inline llama.cpp:2325]: This does not follow the [coding guidelines in the README](https://github.com/ggerganov/llama.cpp#coding-guidelines).

File: llama.cpp [hunk 3]
```diff
@@ -2229,9 +2230,10 @@ static struct ggml_cgraph * llm_build_llama(
     //
     // with the low VRAM option VRAM scratch is disabled in llama_load_model_internal
     // in that case ggml_cuda_assign_buffers has no effect
-    offload_func_t offload_func_nr = llama_nop; // nr = non-repeating
-    offload_func_t offload_func_kq = llama_nop;
-    offload_func_t offload_func_v  = llama_nop;
+    offload_func_t offload_func_nr   = llama_nop; // nr = non-repeating
+    offload_func_t offload_func_kq   = llama_nop;
+    offload_func_t offload_func_v    = llama_nop;
+    offload_func_t offload_func_skip = llama_nop;
```
Comment (JohannesGaessler) [inline llama.cpp:2236]: Why are you defining `offload_func_skip`? If I understand your code correctly it works by discarding the unneeded parts of tensors in the last layer. I don't understand why this would affect the offloading logic, i.e. whether data should be stored in RAM or VRAM.

**Comment by JohannesGaessler** [comment]
I pushed a fix that should have the correct offloading logic [here](https://github.com/JohannesGaessler/llama.cpp/tree/skip-unused-2).

**Comment by SlyEcho** [comment]
If it works all the time everywhere I don't think there needs to be a compile option. But if it stays it should be enabled by default for most users.

**Comment by ochafik** [comment]
@JohannesGaessler thanks so much for looking into it! I still seem to be getting wrong output with that change though (still only broken for `-ngl 1`, see [colab](https://colab.research.google.com/gist/ochafik/664f78524d8650a7b71b322d1b55d98d/llama-cpp-skip-unused-logits-computations.ipynb#scrollTo=uNU7TtHEIxR9&line=1&uniqifier=1) 😓).

File: llama.cpp (hunk 2 above)
Comment (ochafik) [reply llama.cpp]: Thanks, fixed.

File: llama.cpp (hunk 3 above)
Comment (ochafik) [reply llama.cpp]: I was hoping `offload_func_skip` being distinct would help things, but there seems to be some entangling anyway.

Not sure I understand the interplay between `ggml_cuda_compute_forward` and the CPU fallback route, but I've noticed the former skips GGML_OP_VIEW if its input isn't on the GPU, so I've now tried to make sure the offload of `cur` and `inpSA`'s subviews matches whatever backend they currently were using the following defensive code, and... it seems to fix the issue:

```c++
auto cur_on_gpu = cur->backend == GGML_BACKEND_GPU;
cur   = ggml_view_2d(ctx0, cur,   n_embd, 1,   cur->nb[1], (N - 1)*ggml_element_size(cur)*n_embd);
if (cur_on_gpu) ggml_cuda_assign_buffers_no_alloc(cur);

auto inpSA_on_gpu = inpSA->backend == GGML_BACKEND_GPU;
inpSA = ggml_view_2d(ctx0, inpSA, n_embd, 1, inpSA->nb[1], (N - 1)*ggml_element_size(inpSA)*n_embd);
if (inpSA_on_gpu) ggml_cuda_assign_buffers_no_alloc(inpSA);
```

The issue being when `-ngl 1`, `i_gpu_start = n_layer - 1` so the last layer is the only one that offloads most of its tensors (as per the `if (il >= i_gpu_start) { offload_func = ggml_cuda_assign_buffers_no_alloc; }` block).

Another simpler fix (pushed on this PR) that also seems to work is to offload the view inputs (but that may be offloading more than intended?):

```c++
offload_func(cur);
cur   = ggml_view_2d(ctx0, cur,   n_embd, 1,   cur->nb[1], (N - 1)*ggml_element_size(cur)*n_embd);
offload_func(cur);

offload_func(inpSA);
inpSA = ggml_view_2d(ctx0, inpSA, n_embd, 1, inpSA->nb[1], (N - 1)*ggml_element_size(inpSA)*n_embd);
offload_func(inpSA);
```

Will try to find nicer ways to fix this (wondering if the CPU view fallback in a mostly GPU context is behaving as expected?), suggestions welcome 🙂

**Comment by JohannesGaessler** [comment]
Good catch with -ngl 1, it seems that only that exact value was broken.

I pushed more changes to my own repository. The changes are an extra check to the CUDA code to let you safely offload the same tensor multiple times (which I think currently can happen) and a removal of the first call to `offload_func`: since the preceding tensor is in the same layer it should always already be offloaded.

File: llama.cpp (hunk 3 above)
Comment (JohannesGaessler) [reply llama.cpp]: >Not sure I understand the interplay between ggml_cuda_compute_forward and the CPU fallback route

The decision for whether or not CUDA should be used is based on data location. If the data of any of the tensors is in VRAM then all of the calculations are done on the GPU and the data is copied to/from VRAM as necessary based on backend. My preferred design would have been to add a state to the ggml context that sets the backend for all newly created tensors. But unfortunately this way vetoed by Georgi who wanted to avoid GPU offloading logic in ggml. So the current design is kind of cancerous where you have to manually offload each tensor. For tensors that don't actually do any computations this is very awkward because there is no step where the data would be carried over if necessary. Maybe I should add an explicit check for those tensors to ensure that the backends are consistent.

**Comment by ochafik** [comment]
@JohannesGaessler thanks a lot, I've merged your changes! (sorry for the delay, faced some mysterious segfaults on Colab, seems to have been fixed by merging master again 😅)

File: Makefile [hunk 4]
```diff
@@ -441,6 +441,15 @@ k_quants.o: k_quants.c k_quants.h
 	$(CC) $(CFLAGS) -c $< -o $@
 endif # LLAMA_NO_K_QUANTS
 
+ifndef LLAMA_NO_SKIP_UNUSED_LOGITS
+	CFLAGS   += -DLLAMA_SKIP_UNUSED_LOGITS
```
Comment (Cebtenzzre) [inline Makefile:445]: You can't put indented assignments below a build rule like this. The ifdef doesn't make a difference syntactically; this line is treated as part of the k-quants build.

Comment (ochafik) [reply Makefile]: Ah thanks, I'd overlooked https://github.com/ggerganov/llama.cpp/commit/4dcd47d71df8ca4edcc31302744bd93f0c31298e (and hadn't merged it well), fixed.

**Comment by ochafik** [comment]
Updated the description w/ CPU speedup figures for TinyLlamas models ([15-25% faster](https://gist.github.com/ochafik/d75ef799d7288ddd3238bcb2b7631a66))
